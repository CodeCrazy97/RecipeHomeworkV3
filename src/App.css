.page {
  text-align: center;
  margin: 20px;
}

.split {
  height: 100%;
  width: 49.5%;
  position: fixed;
  z-index: 1;
  top: 0;
  overflow-x: hidden;
  padding-top: 20px;
  padding-bottom: 20px;
  text-align: center;
  border: 5px solid brown;
}

/* Control the left side */
.left {
  left: 0;
  padding: 5px;
  background-color: lightsalmon;
}

/* Control the right side */
.right {
  right: 0;
  padding: 5px;
  background-color: palegreen;
}

.bottom-right-button {
  margin: 20px;
  background-color: yellow;
  cursor: pointer;
}

.bottom-left-button {
  margin: 20px;
  background-color: yellow;
  cursor: pointer;
}

.small-image {
  left: 0px;
  position: relative;
  width: 150px;
}

.page-right {
  text-align: left;
  margin: 5px;
  right: 0;
  position: absolute;
}

.page-left {
  text-align: left;
  margin: 5px;
  right: 70%;
  position: absolute;
}

.homepage-link {
  cursor: pointer;
}

.interactions {
  text-align: center;
}

.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
}

.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 0;
  font-size: inherit;
  cursor: pointer;
}

.button-active {
  border-radius: 0;
  border-bottom: 1px solid #38bb6c;
}

.bottom-right-button {
  position: relative;
  right: 20;
  bottom: 20;
}

.bottom-left-button {
  position: relative;
  right: 0;
  bottom: 0;
}
