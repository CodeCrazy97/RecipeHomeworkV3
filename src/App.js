import "./App.css";
import React, { Component } from "react";

// displayOtherRecipe: called when Previous or Next buttons is clicked. Determines which recipe to display next.
function displayOtherRecipe(index, recipeCount, prevButtonClicked) {
  // If it was the "Previous Button" click that caused this function to be called, then display the previous recipe.
  // Otherwise, it was the "Next Button" that caused this function to be called - display the next recipe.
  if (prevButtonClicked) {
    if (index == 0) {
      // Reached the beginning of the recipe list - go to the end.
      return recipeCount - 1;
    } else {
      return index - 1;
    }
  } else {
    if (index == recipeCount - 1) {
      // Reached the end of the recipe list - loop back to the beginning.
      return 0;
    } else {
      return index + 1;
    }
  }
}

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      indexTest: 0,
      recipes: [{}],
      imageClick: this.imageClick,
      previousButtonClick: this.previousButtonClick,
      nextButtonClick: this.nextButtonClick
    };

    this.selectedIndex = 0; // Keeps track of which index into the recipes array was clicked. Default is first recipe in the list.
    this.imageClick = this.imageClick.bind(this);
    this.previousButtonClick = this.previousButtonClick.bind(this);
    this.nextButtonClick = this.nextButtonClick.bind(this);
  }

  componentDidMount() {
    const url = "".concat(process.env.REACT_APP_PATH_BASE, process.env.REACT_APP_JSON_EXTENSION);

    fetch(url)
      .then(response => response.json())
      .then(data => this.setState({ recipes: data }));
  }

  // Called when Previous button is clicked.
  previousButtonClick = oldIndex => {
    this.selectedIndex = displayOtherRecipe(
      oldIndex,
      this.state.recipes.length,
      true
    );

    // Set the currently viewed recipe to the one that was clicked.
    this.setState({ selectedIndex: this.selectedIndex });
  };

  // Called when Next button is clicked.
  nextButtonClick = oldIndex => {
    this.selectedIndex = displayOtherRecipe(
      oldIndex,
      this.state.recipes.length,
      false
    );
    this.setState({ selectedIndex: this.selectedIndex });
  };

  // imageClick: called when a recipe picture, title, or summary is clicked
  // Will take user to page that displays the details of the recipe.
  imageClick = index => {
    this.selectedIndex = index;
    this.setState({ selectedIndex: this.selectedIndex });
  };

  render() {
    const numberOfRecipes = Object.keys(this.state.recipes[0]).length;
    let rc;
    if (numberOfRecipes > 0) {
      rc = (
        <div>
          <RecipeCollection list={this.state.recipes} state={this.state} />
          <Recipe index={this.selectedIndex} state={this.state} />
        </div>
      );
    } else {
      rc = (
        <div className="empty-recipes-message">
          <br />
          <br />
          <br />
          <br />
          <br />
          There are no recipes to view! <br />
          You might not be connected to the Internet.
        </div>
      );
    }

    return (
      <div>
        {(document.title = process.env.REACT_APP_TAB_TITLE)}
        {rc}
      </div>
    );
  }
}

// createMarkup: returns the data in HTML format. Used with JSON that contains HTML elements.
// data - the JSON text that (may) contain HTML elements.
function createMarkup(data) {
  return { __html: data };
}

// Recipe: shows details about the first recipe in the list or a recipe that has been clicked on.
const Recipe = ({ index, state }) => (
  <div className="split-right">
    <span>
      <img src={process.env.REACT_APP_PATH_BASE + state.recipes[index].field_images} />
    </span>
    <span>
      <h1> {state.recipes[index].title} </h1>
    </span>
    <span>
      <h2> Ingredients: {state.recipes[index].field_ingredients} </h2>
    </span>
    <span>
      <h3 dangerouslySetInnerHTML={createMarkup(state.recipes[index].body)} />
    </span>
    <span>
      <input
        className="bottom-left-button"
        type="button"
        value="Previous"
        onClick={() => state.previousButtonClick(index)}
      />
      <input
        className="bottom-right-button"
        type="button"
        value="Next"
        onClick={() => state.nextButtonClick(index)}
      />
    </span>
  </div>
);

// RecipeCollection: displays a brief summary of the recipes on the left-hand side of the screen.
const RecipeCollection = ({ list, state }) => (
  <div className="split-left">
    {list.map((item, index) => (
      <div key={item.objectID}>
        <a
          style={{ cursor: "pointer" }}
          type="button"
          className="card-link"
          onClick={() => state.imageClick(index)}
        >
          <span>
            <img className="small-image" src={process.env.REACT_APP_PATH_BASE + item.field_images} />
          </span>
          <span className="text-right">
            <h3> {item.title} </h3>
            <h4 dangerouslySetInnerHTML={createMarkup(item.field_summary)} />
          </span>
        </a>
        <br />
        <br />
      </div>
    ))}
  </div>
);

export default App;

export {
  App,
  Recipe,
  RecipeCollection, 
  displayOtherRecipe
};

/*
,
  previousButtonClick,
  nextButtonClick,
  ,
  imageClick
*/