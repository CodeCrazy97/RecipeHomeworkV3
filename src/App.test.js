import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import App, {
  previousButtonClick,
  nextButtonClick,
  imageClick,
  displayOtherRecipe,
  Recipe,
  RecipeCollection
} from "./App";
import Adapter from "enzyme-adapter-react-16";
import { Enzyme, render, mount, shallow, configure } from "enzyme";
import { URL } from "url";
import { stat } from "fs";

const props = {
  index: 0,
  state: {
    recipes: [
      {
        title: "Tomato and olive focaccia (TM)",
        body:
          "\n<p> PREPARATION TIME :&nbsp;20<br>COOKING TIME :&nbsp;45<br> SERVES :&nbsp;TODO &lt;<br>DIFFICULTY: EASY<br></p>\n\n\n\n<p> Add the water and the flour to the bowl 30 sec / Speed 5.,Leave for 1 hour to autolyse.,Add the yeast and salt to separate sides of the bowl 10 seconds / Speed 7.,Then 2 minutes / Knead.,Remove the dough from the bowl and form a ball.,Place the dough into a bowl and pour olive oil over the dough to make sure it is well coated.,Cover and leave to rest overnight in the fridge.,The next day, remove from the fridge and bring to room temperature.,Pour enough olive oil to cover the base of a roasting pan about 20 x 30cm.,Put the dough in the centre of the pan and let it rest, covered, for 30 minutes.,Turn the dough in the oil (it must be nice and oily) and spread it with the tip of your fingers from the centre.,Sprinkle with tomatoes pieces or cherry tomatoes, cut in half and push well into the dough with the olives.,Sprinkle with oregano, salt and more oil.,Allow to rest until doubled in size.,Cook on the bottom rack of a preheated oven for 25 minutes at 200° C (400° F - gas 6), [fan oven 180° C],Allow to cool on a wire rack. </p>\n",
        field_images:
          "/sites/default/files/styles/medium/public/2019-10/542904-pxhere.jpg?itok=kKS71aFe, /sites/default/files/styles/medium/public/2019-10/vigo%20the%20cat%20wallpaper.jpg?itok=C7xZGmBE",
        view_node: "/node/8",
        field_ingredients:
          "ingredient 1, ingredient 2, ingredient 3, ingredient 4",
        field_summary:
          "baked flatbread type dish that is easy for even terrible cooks to make (now with 15% less housefire)<br />\r\n"
      },
      {
        title: "Majorcan vegetable bake",
        body:
          "\n<p> PREPARATION TIME : 15<br>  COOKING TIME : 25<br> SERVES : TODO &lt;<br> DIFFICULTY : EASY</p>\n\n\n\n<p> Add 1.5 tablespoons of the oil to a pan and sizzle the garlic for a minute and then add the tomatoes, oregano, season and simmer for 10 minutes to reduce a little.,Toss the aubergine, pepper and onion in the remaining oil, season and arrange in a baking tray and oven roast for 10 minutes.,Remove from the oven and arrange the potatoes and vegetables in layers.,Pour over the tomato sauce, add a grind of black pepper and a glug of olive oil.,Cover with tin foil and bake for 1 hour, removing the tin-foil half way through.<br></p>\n",
        field_images:
          "/sites/default/files/styles/medium/public/2019-10/img_9212-683x1024.jpg?itok=yDqHlTFF",
        view_node: "/node/7",
        field_ingredients:
          "400g can peeled plum tomatoes - chopped, 1 tablespoon fresh oregano or marjoram - chopped, 3 tablespoon olive oil plus a little more to dress the dish with, 6 fat Cloves of garlic (or to taste), half peeled and crushed, the remaining half peeled and roughly chopped, 1 aubergine, 1 sweet bell pepper of any colour, 1 large onion, peeled and sliced, 650g potatoes, sea salt and freshly ground black pepper",
        field_summary: ""
      }
    ]
  }
};

configure({ adapter: new Adapter() });

describe("recipe homework", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
  });

  test("matches the snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // The RecipeCollection component will render, as the recipes list contains data.
  it("contains split-left component", () => {
    const element = render(<RecipeCollection list={props.state.recipes} />);
    expect(element.hasClass("split-left"));
  });

  // The RecipeCollection component won't render, as the recipes list is empty.
  it("does not contain split-left component", () => {
    const element = render(<RecipeCollection list={props.state.recipes} />);
    expect(!element.hasClass("split-left"));
  });

  it("returns index of first recipe", () => {
    expect(displayOtherRecipe(3, 4, false)).toEqual(0);
  });

  it("returns index of last recipe", () => {
    expect(displayOtherRecipe(0, 6, true)).toEqual(5);
  });

  it("returns index of next recipe", () => {
    expect(displayOtherRecipe(2, 4, false)).toEqual(3);
  });

  it("returns index of previous recipe", () => {
    expect(displayOtherRecipe(3, 6, true)).toEqual(2);
  });

  it("test click event", () => {
    const mockCallBack = jest.fn();

    const button = shallow(
      <Recipe index={0} state={(props.state.recipes, mockCallBack)}>
        Previous
      </Recipe>
    );
    button.find("bottom-left-button").simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
