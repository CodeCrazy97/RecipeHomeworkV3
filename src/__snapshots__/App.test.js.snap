// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipe homework matches the snapshot  1`] = `<div />`;

exports[`recipe homework matches the snapshot 1`] = `<div />`;
